File: sshlib/src/test/java/org/connectbot/sshlib/struct/CaptureTest.java
/*
 * Copyright 2019 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.sshlib.struct;

import io.kaitai.struct.ByteBufferKaitaiStream;
import io.kaitai.struct.KaitaiStream;
import io.kaitai.struct.KaitaiStruct;
import org.junit.Test;

import java.io.DataInputStream;
import java.io.InputStream;
import java.math.BigInteger;
import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

public class CaptureTest {
    @Test
    public void serverToClient() throws Exception {
        InputStream is = CaptureTest.class.getResourceAsStream("server-to-client.cap");
        DataInputStream dis = new DataInputStream(is);
        byte[] bytes = new byte[dis.available()];
        dis.readFully(bytes);
        ByteBufferKaitaiStream bb = new ByteBufferKaitaiStream(bytes);
        IdBanner banner = new IdBanner(bb);
        banner._read();
        System.out.println("Banner: " + banner.protoVersion());

        int seqNum = 1;
        while (printPlain(bb) != SshEnums.MessageType.SSH_MSG_NEWKEYS) { seqNum++; }
        while (!bb.isEof()) {
            printEnc(bb, seqNum++);
        }
    }

    private SshEnums.MessageType printPlain(ByteBufferKaitaiStream bb) {
        UnencryptedPacket msg = new UnencryptedPacket(bb);
        msg._read();
        System.out.print("unencrypted msg: ");
        switch (msg.payload().messageType()) {
            case SSH_MSG_KEXINIT:
                System.out.println(msg.payload().messageType());
                SshMsgKexinit init = (SshMsgKexinit) msg.payload().body();
                System.out.print("serverHostKeyAlgorithms: ");
                System.out.println(init.serverHostKeyAlgorithms());
                break;

            case SSH_MSG_KEX_METHOD_SPECIFIC_30:
            case SSH_MSG_KEX_METHOD_SPECIFIC_31:
            case SSH_MSG_KEX_METHOD_SPECIFIC_32:
            case SSH_MSG_KEX_METHOD_SPECIFIC_33:
            case SSH_MSG_KEX_METHOD_SPECIFIC_34:
            case SSH_MSG_KEX_METHOD_SPECIFIC_35:
            case SSH_MSG_KEX_METHOD_SPECIFIC_36:
            case SSH_MSG_KEX_METHOD_SPECIFIC_37:
            case SSH_MSG_KEX_METHOD_SPECIFIC_38:
            case SSH_MSG_KEX_METHOD_SPECIFIC_39:
            case SSH_MSG_KEX_METHOD_SPECIFIC_40:
            case SSH_MSG_KEX_METHOD_SPECIFIC_41:
            case SSH_MSG_KEX_METHOD_SPECIFIC_42:
            case SSH_MSG_KEX_METHOD_SPECIFIC_43:
            case SSH_MSG_KEX_METHOD_SPECIFIC_44:
            case SSH_MSG_KEX_METHOD_SPECIFIC_45:
            case SSH_MSG_KEX_METHOD_SPECIFIC_46:
            case SSH_MSG_KEX_METHOD_SPECIFIC_47:
            case SSH_MSG_KEX_METHOD_SPECIFIC_48:
            case SSH_MSG_KEX_METHOD_SPECIFIC_49:
                kexdhPrint(msg);
                break;
            default:
                System.out.println(msg.payload().messageType());
                break;
        }
        return msg.payload().messageType();
    }

    private void kexdhPrint(UnencryptedPacket msg) {
        KexdhPayload kexdhPayload = new KexdhPayload(
                new ByteBufferKaitaiStream(msg._raw_payload()), msg);
        kexdhPayload._read();
        System.out.println(kexdhPayload.messageType());
        switch (kexdhPayload.messageType()) {
            case SSH_MSG_KEXDH_INIT:
                System.out.print("e: ");
                SshMsgKexdhInit dhinit = (SshMsgKexdhInit) kexdhPayload.body();
                System.out.println(new BigInteger(1, dhinit.e().body()).toString(16));
                break;
            case SSH_MSG_KEXDH_REPLY:
                SshMsgKexdhReply dhreply = (SshMsgKexdhReply) kexdhPayload.body();
                System.out.print("server_key: ");
                System.out.println(new String(dhreply.serverKey().data()));
                System.out.print("f: ");
                System.out.println(new BigInteger(1, dhreply.f().body()).toString(16));
                break;
        }
    }

    private static final String MAC_ALGO = "HmacSHA256";

    /*
     * hash
     * 0000: 9e f5 c8 c3 7c 33 75 d9 46 65 53 a3 e9 44 16 22  ....|3u.FeS..D."
     * 0016: 14 b8 1b b8 0a 1e 39 32 b2 89 96 57 14 13 7b c1  ......92...W..{.
     * key 'A'== key (initial IV client-to-server)
     * 0000: 76 20 6e 50 ff 17 43 9c 50 5d 2b 1a 96 52 42 8c  v nP..C.P]+..RB.
     * 0016: 55 af d8 31 0f 86 52 75 d5 fc 9f 70 ab d2 ce 45  U..1..Ru...p...E
     * key 'B'== key (initial IV server-to-client)
     * 0000: 4a d0 83 35 66 c6 2e da 2b f4 5b 44 79 da cb c2  J..5f...+.[Dy...
     * 0016: 3f b7 ac 4a 2d 5d b8 73 55 fa 81 1b 7b 46 fc c4  ?..J-].sU...{F..
     * key 'C'== key (encryption key client-to-server)
     * 0000: 9d 53 e1 10 b6 62 41 2c 06 a5 b7 6d ba 98 b1 23  .S...bA,...m...#
     * 0016: 7d 37 74 4e 15 9c f7 fd 59 e8 4e 12 aa aa f0 40  }7tN....Y.N....@
     * key 'D'== key (encryption key server-to-client)
     * 0000: f0 6f c5 74 24 38 6c 6c 53 53 a0 2b a2 c0 45 82  .o.t$8llSS.+..E.
     * 0016: de 61 50 f2 41 b1 f8 4f 97 a5 d2 22 e4 d6 96 eb  .aP.A..O..."....
     * key 'E'== key (mac key client-to-server)
     * 0000: 26 4f 54 6a c7 64 f9 96 2c 62 08 c0 6d ce 50 93  &OTj.d..,b..m.P.
     * 0016: 26 ae d9 31 7e 8c 2b b6 11 8c b1 f9 2b 32 14 24  &..1~.+.....+2.$
     * key 'F'== key (mac key server-to-client)
     * 0000: e0 03 a2 02 7d 34 06 f0 6c 9f e4 45 3d f0 e1 e0  ....}4..l..E=...
     * 0016: b6 fb 21 0d 6d a6 fc 5a e4 86 61 ac 00 33 c4 ce  ..!.m..Z..a..3..
     */
    private static final byte[] initialIv_StoC = new byte[] {
(byte) 0x4a, (byte) 0xd0, (byte) 0x83, (byte) 0x35, (byte) 0x66, (byte) 0xc6, (byte) 0x2e, (byte) 0xda, (byte) 0x2b, (byte) 0xf4, (byte) 0x5b, (byte) 0x44, (byte) 0x79, (byte) 0xda, (byte) 0xcb, (byte) 0xc2,
/* (byte) 0x3f, (byte) 0xb7, (byte) 0xac, (byte) 0x4a, (byte) 0x2d, (byte) 0x5d, (byte) 0xb8, (byte) 0x73, (byte) 0x55, (byte) 0xfa, (byte) 0x81, (byte) 0x1b, (byte) 0x7b, (byte) 0x46, (byte) 0xfc, (byte) 0xc4, */
    };
    private static final byte[] cipherKey_StoC = new byte[] {
(byte) 0xf0, (byte) 0x6f, (byte) 0xc5, (byte) 0x74, (byte) 0x24, (byte) 0x38, (byte) 0x6c, (byte) 0x6c, (byte) 0x53, (byte) 0x53, (byte) 0xa0, (byte) 0x2b, (byte) 0xa2, (byte) 0xc0, (byte) 0x45, (byte) 0x82,
/*(byte) 0xde, (byte) 0x61, (byte) 0x50, (byte) 0xf2, (byte) 0x41, (byte) 0xb1, (byte) 0xf8, (byte) 0x4f, (byte) 0x97, (byte) 0xa5, (byte) 0xd2, (byte) 0x22, (byte) 0xe4, (byte) 0xd6, (byte) 0x96, (byte) 0xeb,*/
    };
    private static final SecretKeySpec CIPHER_KEY = new SecretKeySpec(cipherKey_StoC, "AES");

    private static final byte[] hmacKey_StoC = new byte[] {
        (byte) 0xe0, (byte) 0x03, (byte) 0xa2, (byte) 0x02, (byte) 0x7d, (byte) 0x34, (byte) 0x06, (byte) 0xf0, (byte) 0x6c, (byte) 0x9f, (byte) 0xe4, (byte) 0x45, (byte) 0x3d, (byte) 0xf0, (byte) 0xe1, (byte) 0xe0,
        (byte) 0xb6, (byte) 0xfb, (byte) 0x21, (byte) 0x0d, (byte) 0x6d, (byte) 0xa6, (byte) 0xfc, (byte) 0x5a, (byte) 0xe4, (byte) 0x86, (byte) 0x61, (byte) 0xac, (byte) 0x00, (byte) 0x33, (byte) 0xc4, (byte) 0xce,
    };
    private static final SecretKeySpec MAC_KEY = new SecretKeySpec(hmacKey_StoC, MAC_ALGO);

    private byte[] iv = initialIv_StoC.clone();

    private byte[] toByteArray(KaitaiStruct.ReadWrite struct) throws Exception {
	struct._check();
        KaitaiStream io = new ByteBufferKaitaiStream(1024 * 16);
        struct._write(io);
        long size = io.pos();
        io.seek(0);
        return io.readBytes(size);
    }

    private void printEnc(ByteBufferKaitaiStream bb, int seqNum) throws Exception {
        EncryptedPacket msg = new EncryptedPacket(bb, 32);
        msg._read();
        System.out.print("encrypted size: ");
        System.out.println(msg.lenEncryptedPayload());
        Mac mac = Mac.getInstance(MAC_ALGO);
        mac.init(MAC_KEY);

        EtmMac macInput = new EtmMac();
        macInput.setLenEncryptedPacket(msg.lenEncryptedPayload());
        macInput.setEncryptedPacket(msg.encryptedPayload());
        macInput.setSequenceNumber(seqNum);
        byte[] tag = mac.doFinal(toByteArray(macInput));

        System.out.println("calculated tag: " + new BigInteger(1, tag).toString(16));
        System.out.println("    actual tag: " + new BigInteger(1, msg.mac()).toString(16));

        Cipher c = Cipher.getInstance("AES/CTR/NoPadding");
        IvParameterSpec ivSpec = new IvParameterSpec(iv);

        c.init(Cipher.DECRYPT_MODE, CIPHER_KEY, ivSpec);
        byte[] decryptedBytes = c.doFinal(msg.encryptedPayload());

        System.out.println("decrypted: " + new BigInteger(1, decryptedBytes).toString(16));

        incrementIv(iv, decryptedBytes.length / c.getBlockSize());

        ByteBufferKaitaiStream dbytes = new ByteBufferKaitaiStream(decryptedBytes);
        EncryptedPacket.DecryptedPacket dmsg = new EncryptedPacket.DecryptedPacket(dbytes, msg);
        dmsg._read();
        System.out.print("decrypted msg: ");
        System.out.println(dmsg.payload().messageType());
        switch (dmsg.payload().messageType()) {
            case SSH_MSG_CHANNEL_DATA:
                SshMsgChannelData data = (SshMsgChannelData) dmsg.payload().body();
                System.out.print("channel: ");
                System.out.println(data.recipientChannel());
                System.out.print("data: ");
                System.out.println(new String(data.data().data()));
                break;
        }
    }

    private void incrementIv(byte[] iv, int numBlocks) {
        int carry = numBlocks;
        for (int i = iv.length - 1; i > 0; i--) {
            int sum = (iv[i] & 0xFF) + carry;
            iv[i] = (byte) sum;
            carry = sum >> 8;
            if (carry == 0)
                break;
        }
    }
}

File: sshlib/src/test/kotlin/org/connectbot/sshlib/protocol/CaptureTest.kt
/*
 * ConnectBot SSH Library
 * Copyright 2025 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.connectbot.sshlib.protocol

import io.kaitai.struct.ByteBufferKaitaiStream
import kotlinx.coroutines.runBlocking
import org.connectbot.sshlib.transport.ByteArrayTransport
import org.connectbot.sshlib.transport.PacketIO
import org.junit.jupiter.api.Test
import java.math.BigInteger
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * Tests that parse captured SSH protocol data.
 *
 * These tests demonstrate:
 * 1. Using ByteArrayTransport to feed captured bytes into the library
 * 2. Parsing unencrypted SSH packets via PacketIO
 * 3. Manual decryption of ETM (Encrypt-then-MAC) encrypted packets
 *
 * The captured data uses ETM mode which is different from the standard
 * encrypt-and-MAC mode currently in PacketIO, so encrypted packets are
 * parsed manually using the derived keys from the capture.
 */
class CaptureTest {

    /**
     * Parse the server-to-client capture file using ByteArrayTransport and PacketIO.
     */
    @Test
    fun `parse unencrypted packets via PacketIO`() = runBlocking {
        val captureData = loadCaptureFile("server-to-client.cap")
        val transport = ByteArrayTransport(captureData)
        val packetIO = PacketIO(transport)

        // Read banner
        val banner = packetIO.readBanner()
        println("Banner: ${banner.protoVersion().trimEnd('\r', '\n')}")
        assertTrue(banner.protoVersion().startsWith("2.0"))

        // Read unencrypted packets until NEWKEYS
        var packetCount = 0
        while (true) {
            val packet = packetIO.readPacket()
            packetCount++
            println("Packet $packetCount: ${packet.messageType()}")

            when (packet.messageType()) {
                SshEnums.MessageType.SSH_MSG_KEXINIT -> {
                    val kexInit = packet.body() as SshMsgKexinit
                    println("  KEX algorithms: ${kexInit.kexAlgorithms().entries().data()}")
                    println("  Host key algorithms: ${kexInit.serverHostKeyAlgorithms().entries().data()}")
                }

                SshEnums.MessageType.SSH_MSG_NEWKEYS -> {
                    println("  NEWKEYS received - switching to encrypted mode")
                    break
                }

                else -> {
                    // KEX method specific messages need special handling
                    if (packet.messageType().id() in 30..49) {
                        parseKexMethodSpecific(packet)
                    }
                }
            }
        }

        println("\nParsed $packetCount unencrypted packets")
        println("Remaining bytes for encrypted packets: ${transport.bytesRemaining}")

        // The rest of the capture is encrypted with ETM mode
        // We'll parse it manually since PacketIO doesn't support ETM yet
        assertTrue(transport.bytesRemaining > 0)
    }

    /**
     * Parse the full capture including encrypted packets using manual decryption.
     *
     * The capture uses ETM (Encrypt-then-MAC) mode with hardcoded derived keys.
     */
    @Test
    fun `parse full capture with manual ETM decryption`() = runBlocking {
        val captureData = loadCaptureFile("server-to-client.cap")
        val transport = ByteArrayTransport(captureData)
        val packetIO = PacketIO(transport)

        // Read banner
        val banner = packetIO.readBanner()
        assertNotNull(banner)

        // Read unencrypted packets until NEWKEYS
        // Sequence numbers: KEXINIT=1, DH_REPLY=2, NEWKEYS=3
        var seqNum = 1
        while (true) {
            val packet = packetIO.readPacket()
            if (packet.messageType() == SshEnums.MessageType.SSH_MSG_NEWKEYS) {
                break
            }
            seqNum++
        }
        println("After NEWKEYS seqNum=$seqNum")

        // Parse encrypted packets manually using ETM mode
        val iv = INITIAL_IV_S_TO_C.clone()
        var encryptedPacketCount = 0

        // For parsing we need to get the remaining bytes as a stream
        val remainingBytes = transport.read(transport.bytesRemaining)
        val encryptedStream = ByteBufferKaitaiStream(remainingBytes)

        while (!encryptedStream.isEof) {
            encryptedPacketCount++

            // Parse using EncryptedPacket structure
            val msg = EncryptedPacket(encryptedStream, MAC_LENGTH.toLong())
            msg._read()

            // Compute MAC (ETM: sequence_number || len_encrypted_packet || encrypted_packet)
            val computedMac = computeHmacSha256(MAC_KEY, seqNum, msg.lenEncryptedPayload(), msg.encryptedPayload())
            println("Packet $encryptedPacketCount seqNum=$seqNum encLen=${msg.lenEncryptedPayload()}")
            println("  actual MAC: ${BigInteger(1, msg.mac()).toString(16)}")
            println("  computed MAC: ${BigInteger(1, computedMac).toString(16)}")
            assertEquals(
                BigInteger(1, msg.mac()).toString(16),
                BigInteger(1, computedMac).toString(16),
                "MAC verification failed for packet $encryptedPacketCount (seqNum=$seqNum)",
            )

            // Decrypt
            val decrypted = decryptAesCtr(CIPHER_KEY, iv, msg.encryptedPayload())

            // Increment IV by number of blocks
            incrementIv(iv, decrypted.size / AES_BLOCK_SIZE)

            // Parse decrypted packet
            val stream = ByteBufferKaitaiStream(decrypted)
            val decryptedPacket = EncryptedPacket.DecryptedPacket(stream, msg)
            decryptedPacket._read()

            println("Encrypted packet $encryptedPacketCount: ${decryptedPacket.payload().messageType()}")

            when (decryptedPacket.payload().messageType()) {
                SshEnums.MessageType.SSH_MSG_CHANNEL_DATA -> {
                    val channelData = decryptedPacket.payload().body() as SshMsgChannelData
                    println("  Channel ${channelData.recipientChannel()}: ${String(channelData.data().data())}")
                }

                else -> {}
            }

            seqNum++
        }

        println("\nParsed $encryptedPacketCount encrypted packets")
        assertTrue(encryptedPacketCount > 0)
    }

    private fun parseKexMethodSpecific(packet: UnencryptedPacket.UnencryptedPayload) {
        val rawPayload = packet._raw_body()
        val messageTypeByte = packet.messageType().id().toByte()
        val fullPayload = byteArrayOf(messageTypeByte) + rawPayload

        val stream = ByteBufferKaitaiStream(fullPayload)
        val kexdhPayload = KexdhPayload(stream)
        kexdhPayload._read()

        println("  KEX method: ${kexdhPayload.messageType()}")

        when (kexdhPayload.messageType()) {
            SshEnums.KexDh.SSH_MSG_KEXDH_INIT -> {
                val init = kexdhPayload.body() as SshMsgKexdhInit
                println("  e: ${BigInteger(1, init.e().body()).toString(16).take(32)}...")
            }

            SshEnums.KexDh.SSH_MSG_KEXDH_REPLY -> {
                val reply = kexdhPayload.body() as SshMsgKexdhReply
                println("  f: ${BigInteger(1, reply.f().body()).toString(16).take(32)}...")
            }

            else -> {}
        }
    }

    private fun loadCaptureFile(name: String): ByteArray {
        val stream = javaClass.getResourceAsStream(name)
            ?: throw IllegalArgumentException("Capture file not found: $name")
        return stream.readBytes()
    }

    companion object {
        private const val AES_BLOCK_SIZE = 16
        private const val MAC_LENGTH = 32

        // Derived keys from the capture (server-to-client direction)
        // These would normally be computed from the key exchange, but for
        // testing with captured data we use the pre-computed values.

        // Key 'B' - Initial IV server-to-client
        private val INITIAL_IV_S_TO_C = byteArrayOf(
            0x4a.toByte(), 0xd0.toByte(), 0x83.toByte(), 0x35.toByte(),
            0x66.toByte(), 0xc6.toByte(), 0x2e.toByte(), 0xda.toByte(),
            0x2b.toByte(), 0xf4.toByte(), 0x5b.toByte(), 0x44.toByte(),
            0x79.toByte(), 0xda.toByte(), 0xcb.toByte(), 0xc2.toByte(),
        )

        // Key 'D' - Encryption key server-to-client
        private val CIPHER_KEY = byteArrayOf(
            0xf0.toByte(), 0x6f.toByte(), 0xc5.toByte(), 0x74.toByte(),
            0x24.toByte(), 0x38.toByte(), 0x6c.toByte(), 0x6c.toByte(),
            0x53.toByte(), 0x53.toByte(), 0xa0.toByte(), 0x2b.toByte(),
            0xa2.toByte(), 0xc0.toByte(), 0x45.toByte(), 0x82.toByte(),
        )

        // Key 'F' - MAC key server-to-client
        private val MAC_KEY = byteArrayOf(
            0xe0.toByte(), 0x03.toByte(), 0xa2.toByte(), 0x02.toByte(),
            0x7d.toByte(), 0x34.toByte(), 0x06.toByte(), 0xf0.toByte(),
            0x6c.toByte(), 0x9f.toByte(), 0xe4.toByte(), 0x45.toByte(),
            0x3d.toByte(), 0xf0.toByte(), 0xe1.toByte(), 0xe0.toByte(),
            0xb6.toByte(), 0xfb.toByte(), 0x21.toByte(), 0x0d.toByte(),
            0x6d.toByte(), 0xa6.toByte(), 0xfc.toByte(), 0x5a.toByte(),
            0xe4.toByte(), 0x86.toByte(), 0x61.toByte(), 0xac.toByte(),
            0x00.toByte(), 0x33.toByte(), 0xc4.toByte(), 0xce.toByte(),
        )

        private fun computeHmacSha256(key: ByteArray, seqNum: Int, encryptedLength: Long, encryptedPayload: ByteArray): ByteArray {
            val etmMac = EtmMac()
            etmMac.setSequenceNumber(seqNum.toLong())
            etmMac.setLenEncryptedPacket(encryptedLength)
            etmMac.setEncryptedPacket(encryptedPayload)

            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(key, "HmacSHA256"))
            return mac.doFinal(etmMac.toByteArray())
        }

        private fun decryptAesCtr(key: ByteArray, iv: ByteArray, data: ByteArray): ByteArray {
            val cipher = Cipher.getInstance("AES/CTR/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), IvParameterSpec(iv))
            return cipher.doFinal(data)
        }

        private fun incrementIv(iv: ByteArray, numBlocks: Int) {
            var carry = numBlocks
            for (i in iv.size - 1 downTo 1) {
                val sum = (iv[i].toInt() and 0xFF) + carry
                iv[i] = sum.toByte()
                carry = sum shr 8
                if (carry == 0) break
            }
        }
    }
}
